lib/_stream_readable.js

```diff
@@ -1070,13 +1072,61 @@ Readable.prototype.wrap = function(stream) {
 };
 
 Readable.prototype[SymbolAsyncIterator] = function() {
+  if (typeof this.read === 'function') {
+    const iter = createAsyncIterator(this);
+    iter.stream = this;
+    return iter;
+  }
+
   if (createReadableStreamAsyncIterator === undefined) {
     createReadableStreamAsyncIterator =
       require('internal/streams/async_iterator');
   }
+
   return createReadableStreamAsyncIterator(this);
 };
 
+async function* createAsyncIterator(stream) {
+  const state = stream._readableState;
+
+  if (state.closeEmitted) {
+    return;
+  }
+
+  const eosPromise = new Promise((resolve, reject) => stream
+    .on('error', (err) => reject(err))
+    .on('end', () => resolve(true))
+    .on('close', () => {
+      if (state.endEmitted) {
+        resolve(true);
+      } else {
+        reject(new ERR_STREAM_PREMATURE_CLOSE());
+      }
+    })
+  ).catch(() => {});
+
+  let error;
+  try {
+    while (true) {
+      const chunk = stream.read();
+      if (chunk !== null) {
+        yield chunk;
+      } else if (await Promise.race([
+        eosPromise,
+        new Promise((resolve) => {
+          stream.once('readable', resolve);
```

Comment:
Isn't this problematic because between the yield and and promise resolving we have a time where no one is subscribed to the event?

I'm not convinced this impl isn't susceptible to timing bugs, especially in streams that are in memory and can  emit readable synchronously multiple times.

Reply:
It shouldn't be a problem since we haven't `read()` from it. When this promise completes it should be possible to `read()` and if/when that fails a new `readable` event should be emitted in the next tick.

Assuming someone else is not doing stuff with the stream... I'll give that scenario some more thought. Though, I think we might have similar problems with the current/previous implemenation as well.

Reply:
Ah, I see what you mean. Good catch.

Reply:
No, actually, that's fine, if chunk !== null then we don't need a readable.

Reply:
> emit readable synchronously multiple times.

readable can't be emitted synchronously, it's always done in a nextTick

Reply:
I'll clarify: this is _probably fine_ for our internal streams, but I'm wondering if userland implementations all have the same (implicit) timing guarantees.

> Assuming someone else is not doing stuff with the stream... I'll give that scenario some more thought. Though, I think we might have similar problems with the current/previous implemenation as well.

I thought the current implementation was the way Matteo did it explicitly for this reason (to not have a timing where you don't have a subscribed listener).

Reply:
@benjamingr PTAL